src/test/java/org/takes/tk/TkRetryTest.java

```diff
@@ -0,0 +1,122 @@
+ * SOFTWARE.
+ */
+
+package org.takes.tk;
+
+import java.io.IOException;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.takes.Request;
+import org.takes.Take;
+import org.takes.rq.RqFake;
+import org.takes.rq.RqMethod;
+import org.takes.rs.RsPrint;
+import org.takes.rs.RsText;
+
+/**
+ * TkRetry can retry till success or retry count is reached.
+ *
+ * @author Hamdi Douss (douss.hamdi@gmail.com)
+ * @version $Id$
+ * @since 0.28.3
+ */
+public final class TkRetryTest {
+
+    /**
+     * TkRetry can work when no IOException.
+     *
+     * @throws Exception if something is wrong
+     */
+    @Test
+    public void worksWithNoException() throws Exception {
+        final String test = "test";
+        MatcherAssert.assertThat(
+                new RsPrint(
+                        new TkRetry(2, 2, new TkText(test))
+                                .act(new RqFake())
+                ).print(),
+                Matchers.containsString(test)
+        );
+    }
+
+    /**
+     * TkRetry can retry when initial take fails with IOException,
+     * till retry count is reached.
+     *
+     * @throws Exception if something is wrong
+     */
+    @Test(expected = IOException.class)
+    public void retriesOnExceptionTillCount() throws Exception {
+        final int count = 3;
+        final int delay = 1000;
+        final Take take = Mockito.mock(Take.class);
+        final Request req = new RqFake(RqMethod.GET);
+        Mockito.when(take.act(Mockito.any(Request.class))).thenThrow(
+                new IOException()
+        );
+        final long startTime = System.currentTimeMillis();
+        try {
+            new TkRetry(count, delay, take).act(req);
+        } catch (final IOException exception) {
+            MatcherAssert.assertThat(
+                new Long(count * delay),
+                Matchers.lessThanOrEqualTo(
+                        System.currentTimeMillis() - startTime
+                )
+            );
+            throw exception;
+        }
+    }
+
+    /**
+     * TkRetry can retry when initial take fails with IOException,
+     * till get successful result.
+     *
+     * @throws Exception if something is wrong
+     */
+    @Test
+    public void retriesOnExceptionTillSuccess() throws Exception {
+        final int count = 3;
+        final int delay = 1000;
+        final String data = "data";
+        final Take take = Mockito.mock(Take.class);
+        Mockito.when(take.act(Mockito.any(Request.class))).thenThrow(
+                new IOException()
+        ).thenReturn(new RsText(data));
+        final long startTime = System.currentTimeMillis();
+        final RsPrint response = new RsPrint(
+            new TkRetry(count, delay, take).act(new RqFake(RqMethod.GET))
+        );
+        MatcherAssert.assertThat(
+            new Long(delay),
+            Matchers.lessThanOrEqualTo(System.currentTimeMillis() - startTime)
```

Comment:
@HDouss let's measure delays using `System.nanoTime` since `currentTimeMillis` is very inaccurate. Also, we should capture the current time as soon as the action is complete on line 113.